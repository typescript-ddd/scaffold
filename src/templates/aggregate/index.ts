export * from "./aggregate-root.generator";
export * from "./aggregate-root.template";
export * from "./aggregate-root.types";
