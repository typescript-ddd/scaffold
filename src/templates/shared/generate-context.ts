import { SourceFile } from "ts-morph";

export type ContextImportDeclaration = { moduleSpecifier: string; namedImports: string[] };

export interface GenerateContext {
    resolveDir(type: "domain" | "infrastructure" | "application" | "presentation", subPath?: string): string;
    fileName: string;
    addImportDeclaration(declaration: ContextImportDeclaration): void;
    getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[];
    getModuleSpecifiers(): string[];
    getNamedImports(moduleSpecifier?: string): string[];    
    writeSourceFileImports: (sourceFile: SourceFile) => void;
    clearImportDeclarations(): void;
    tokenize: (text: string, tokens: { name: string }[]) => string;
}