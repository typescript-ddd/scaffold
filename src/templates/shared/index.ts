export * from "./context-builder";
export * from "./generate-context";