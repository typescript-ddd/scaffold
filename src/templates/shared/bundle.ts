import { GenerateContext, Chunk } from "./generate-context";
import { Template, TemplateOption } from "./template";

export interface Bundle<
  TValues extends { [K in keyof TOptions]: any },
  TOptions extends { [key: string]: TemplateOption }
> extends Omit<Template<TValues, TOptions>, "generate"> {
  generate(value: TValues, context: GenerateContext): Chunk[];
}
