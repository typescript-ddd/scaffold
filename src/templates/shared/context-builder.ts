import { GenerateContext, ContextImportDeclaration } from "./generate-context";
import { arrays } from "../../utils";
import { SourceFile } from "ts-morph";

export class ContextBuilder implements GenerateContext {
  private _fileName: string;
  private _importDeclarations: ContextImportDeclaration[] = [];

  private constructor(
    private rootDir: string = "@/src",
    fileName: string = "example.ts"
  ) {
    this._fileName = fileName;
  }

  resolveDir(
    type: "domain" | "infrastructure" | "application" | "presentation",
    subPath?: string
  ): string {
    switch (type) {
      case "domain":
        return `${this.rootDir}/domain${subPath ? `/${subPath}` : ""}`;
      case "infrastructure":
        return `${this.rootDir}/infrastructure${subPath ? `/${subPath}` : ""}`;
      case "application":
        return `${this.rootDir}/application${subPath ? `/${subPath}` : ""}`;
      case "presentation":
        return `${this.rootDir}/presentation${subPath ? `/${subPath}` : ""}`;
      default:
        throw new Error(`Unknown type ${type} for resolveDir`);
    }
  }

  getImportDeclarations(moduleSpecifier?: string): ContextImportDeclaration[] {
    const modules = this.getModuleSpecifiers();

    const declarations: ContextImportDeclaration[] = [];

    for (const currentModule of modules) {
      const namedImports = this.getNamedImports(currentModule);
      declarations.push({ moduleSpecifier: currentModule, namedImports });
    }

    if (moduleSpecifier) {
      return declarations.filter((x) => x.moduleSpecifier === moduleSpecifier);
    }

    return declarations;
  }

  addImportDeclaration(declaration: ContextImportDeclaration): void {
    this._importDeclarations.push(declaration);
  }

  getModuleSpecifiers(): string[] {
    return arrays.distinct(
      this._importDeclarations.map((x) => x.moduleSpecifier)
    );
  }

  getNamedImports(moduleSpecifier: string): string[] {
    return arrays.distinct(
      this._importDeclarations
        .filter((x) => x.moduleSpecifier === moduleSpecifier)
        .map((x) => x.namedImports)
        .flat()
    );
  }

  writeSourceFileImports(sourceFile: SourceFile) {
    const importDeclarations = this.getImportDeclarations();
    const moduleSpecifiers = this.getModuleSpecifiers();

    const coreImports = importDeclarations.filter(
      (x) => x.moduleSpecifier === "@typescript-ddd/core"
    );

    if (coreImports) {
      sourceFile.addImportDeclaration({
        moduleSpecifier: `@typescript-ddd/core`,
        namedImports: coreImports.map((x) => x.namedImports).flat(),
      });
    }

    // Non core and root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter(
        (x) => x !== "@typescript-ddd/core" && !x.startsWith(this.rootDir)
      )
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }

    // Root imports
    for (const moduleSpecifier of moduleSpecifiers
      .filter((x) => x.startsWith(this.rootDir))
      .sort((a, b) => {
        if (a < b) return -1;
        if (a > b) return 1;
        return 0;
      })) {
      sourceFile.addImportDeclaration({
        moduleSpecifier,
        namedImports: this.getNamedImports(moduleSpecifier),
      });
    }
  }

  clearImportDeclarations(): void {
    this._importDeclarations = [];
  }

  tokenize(text: string, tokens: { name: string }[]): string {
    let output = text;
    tokens.forEach((token) => {
      output = output.replace(`{{${token.name}}}`, "${" + token.name + "}");
    });
    return "`" + output + "`";
  }

  get fileName(): string {
    return this._fileName;
  }

  static build(rootDir: string): ContextBuilder {
    return new ContextBuilder(rootDir);
  }
}
