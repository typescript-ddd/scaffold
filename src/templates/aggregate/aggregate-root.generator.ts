import { Project, Scope, StructureKind } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";
import { AggregateRootTemplateValues } from "./aggregate-root.types";

export const generateAggregateRoot = (
  values: AggregateRootTemplateValues,
  context: GenerateContext
) => {
  const { entityName, properties, trackable = false } = values;
  const name = strings.capitalize(entityName);
  const idName = `${name}Id`;
  const fileName = strings.kebab(entityName);
  const baseName = trackable ? "TrackableAggregateRoot" : "AggregateRoot";

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${fileName}.ts`,
    `
export interface Create${name}Props extends EntityCreateProps {}
export interface Update${name}Props extends EntityUpdateProps<${idName}> {}
export class ${name} extends ${baseName}<${idName}> {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: [
      baseName,
      "EntityCreateProps",
      "EntityUpdateProps",
      trackable ? "UtcDate" : "",
    ],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain") + `/events`,
    namedImports: [
      `${name}CreatedEvent`,
      `${name}UpdatedEvent`,
      `${name}DeletedEvent`,
    ],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain") + `/models`,
    namedImports: [`${name}Id`],
  });

  context.writeSourceFileImports(sourceFile);

  const createInterface = sourceFile.getInterface(`Create${name}Props`);
  createInterface.addProperties(
    properties.map((property) => ({
      name: property.name,
      type: property.valueType,
      hasQuestionToken: property.nullable,
    }))
  );
  createInterface.addJsDoc({
    description: `Represents the properties required to create a ${strings.lower(
      name
    )}.`,
    tags: [{ tagName: "extends", text: "{EntityCreateProps}" }],
  });
  if (trackable) {
    createInterface.addProperties([
      { name: "createdAtUtc", type: "UtcDate", hasQuestionToken: true },
      { name: "updatedAtUtc", type: "UtcDate", hasQuestionToken: true },
    ]);
  }
  const updateInterface = sourceFile.getInterface(`Update${name}Props`);
  updateInterface.addProperties(
    properties.map((property) => ({
      name: property.name,
      type: property.valueType,
      hasQuestionToken: property.nullable,
    }))
  );
  updateInterface.addJsDoc({
    description: `Represents the properties required to update a ${strings.lower(
      name
    )}.`,
    tags: [{ tagName: "extends", text: "{EntityUpdateProps}" }],
  });
  const classDeclaration = sourceFile.getClass(name);
  classDeclaration.addJsDoc({
    description: `Represents a ${strings.lower(name)} aggregate.`,
    tags: [{ tagName: "extends", text: baseName }],
  });
  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: trackable
      ? [
          ...properties.map((property) => ({
            name: property.name,
            type: property.valueType,
            scope: Scope.Private,
            hasQuestionToken: property.nullable,
          })),
          { name: "createdAtUtc", type: "UtcDate", hasQuestionToken: true },
          { name: "updatedAtUtc", type: "UtcDate", hasQuestionToken: true },
          { name: "id", type: `${name}Id`, hasQuestionToken: true },
        ]
      : [
          ...properties.map((property) => ({
            name: property.name,
            type: property.valueType,
            scope: Scope.Private,
            hasQuestionToken: property.nullable,
          })),
          { name: "id", type: `${name}Id`, hasQuestionToken: true },
        ],
  });
  constructorDeclaration.setBodyText((writer) => {
    if (trackable) {
      writer.writeLine(
        `super(id ?? ${idName}.create(), createdAtUtc, updatedAtUtc);`
      );
    } else {
      writer.writeLine(`super(id ?? ${idName}.create());`);
    }
    writer.writeLine(`if (!id) {`);
    writer.writeLine(
      `  this.applyChange(new ${name}CreatedEvent(this.id, this));`
    );
    writer.writeLine(`}`);
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(name)}.`,
    tags: [
      {
        tagName: "param",
        text: `{${name}Id} id - An optional identifier of the ${strings.lower(
          name
        )}.`,
      },
      ...properties.map((property) => ({
        tagName: "param",
        text: `{${property.valueType}${property.nullable ? "?" : ""}} ${
          property.name
        } - The ${property.name} of the ${strings.lower(name)}.`,
      })),
      ...(trackable
        ? [
            {
              tagName: "param",
              text: "{UtcDate} createdAtUtc - An optional creation date.",
            },
            {
              tagName: "param",
              text: "{UtcDate} updatedAtUtc - An optional last update date.",
            },
          ]
        : []),
    ],
  });

  properties.forEach((property) => {
    classDeclaration
      .addMethod({
        name: `get${strings.capitalize(property.name)}`,
        returnType: property.valueType,
        hasQuestionToken: property.nullable,
      })
      .setBodyText(`return this.${property.name};`)
      .addJsDoc({
        description: `Gets the ${property.name} of the ${strings.lower(name)}.`,
        tags: [
          {
            tagName: "returns",
            text: `{${property.valueType}${
              property.nullable ? "?" : ""
            }} - The ${property.name} of the ${strings.lower(name)}.`,
          },
        ],
      });
  });

  classDeclaration
    .addMethod({
      name: "update",
      returnType: "void",
      parameters: [{ name: "props", type: `Update${name}Props` }],
      docs: [
        {
          description: `Update the ${strings.natural(name)}.`,
          kind: StructureKind.JSDoc,
          tags: [
            {
              tagName: "param",
              text: `{Update${name}Props} props - The properties to update.`,
            },
            {
              tagName: "returns",
              text: "void",
            },
          ],
        },
      ],
    })
    .setBodyText((writer) => {
      writer.writeLine(`// TODO: validate props`);
      properties.forEach((property) => {
        writer.writeLine(`this.${property.name} = props.${property.name};`);
      });
      if (trackable) {
        writer.writeLine(`this.updatedAtUtc = UtcDate.now();`);
      }
      writer.writeLine(
        `this.applyChange(new ${name}UpdatedEvent(this.id, this));`
      );
    });

  classDeclaration
    .addMethod({
      name: "delete",
      returnType: "void",
    })
    .setBodyText((writer) => {
      writer.writeLine(
        `this.applyChange(new ${name}DeletedEvent(this.id, this));`
      );
    });

  classDeclaration
    .addMethod({
      name: "create",
      isStatic: true,
      parameters: [
        { name: "props", type: `Create${name}Props` },
        {
          name: "id",
          hasQuestionToken: true,
          type: `${name}Id`,
        },
      ],
      returnType: name,
    })
    .setBodyText((writer) => {
      writer.writeLine(`const ${strings.lower(name)} = new ${name}(`);
      properties.forEach((property, index) => {
        writer.writeLine(`  props.${property.name},`);
      });
      writer.writeLine(`  id`);
      writer.writeLine(`);`);
      writer.writeLine(`return ${strings.lower(name)};`);
    });

  let output = sourceFile.getFullText();

  return output;
};
