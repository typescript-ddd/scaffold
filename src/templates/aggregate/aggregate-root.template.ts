import { GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateAggregateRoot } from "./aggregate-root.generator";
import {
  AggregateRootTemplateValues,
  AggregateRootTemplateOptions,
} from "./aggregate-root.types";

export class AggregateRootTemplate
  implements
    Template<AggregateRootTemplateValues, AggregateRootTemplateOptions>
{
  readonly name = "Aggregate Root";
  readonly description = "Generates an aggregate root";
  readonly options: AggregateRootTemplateOptions = {
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the aggregate",
      optionType: "string",
    },
    properties: {
      name: "properties",
      label: "Properties",
      description: "The properties of the aggregate",
      optionType: "properties",
    },
    trackable: {
      name: "trackable",
      label: "Trackable",
      description: "Whether the aggregate is trackable",
      optionType: "boolean",
    },
  };
  readonly defaultValues = {
    entityName: "User",
    properties: [{ name: "name", valueType: "string" }],
    trackable: false,
  };
  generate(
    values: AggregateRootTemplateValues,
    context: GenerateContext
  ): string {
    return generateAggregateRoot(values, context);
  }
}
