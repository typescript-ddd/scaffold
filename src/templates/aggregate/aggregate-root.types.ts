import { TemplateBooleanOption, TemplatePropertiesOption, TemplatePropertyValue, TemplateStringOption } from "../shared/template";

export type AggregateRootTemplateValues = {
  entityName: string;
  properties: TemplatePropertyValue[];
  trackable: boolean;
};

export type AggregateRootTemplateOptions = {
  entityName: TemplateStringOption;
  properties: TemplatePropertiesOption;
  trackable: TemplateBooleanOption;
};
