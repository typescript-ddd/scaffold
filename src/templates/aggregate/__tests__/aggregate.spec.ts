import { ContextBuilder, GenerateContext } from "../../shared";
import { generateAggregate } from "../aggregate";

describe("Aggregate", () => {
  let context: GenerateContext;

  beforeEach(() => {
    context = ContextBuilder.build("@src");
  });

  it("Should render the aggregate component", () => {
    const output = generateAggregate(
      {
        entityName: "User",
        properties: [{ name: "name", type: "string" }],
        trackable: false,
      },
      context
    );

    expect(output).toMatchSnapshot();
  });

  it("Should render the trackable aggregate component", () => {
    const output = generateAggregate(
      {
        entityName: "User",
        properties: [{ name: "name", type: "string" }],
        trackable: true,
      },
      context
    );

    expect(output).toMatchSnapshot();
  });
});
