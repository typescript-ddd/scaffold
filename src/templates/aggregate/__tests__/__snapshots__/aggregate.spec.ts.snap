// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Aggregate Should render the aggregate component 1`] = `
"import { AggregateRoot, EntityCreateProps, EntityUpdateProps  } from "@typescript-ddd/core";
import { UserCreatedEvent, UserUpdatedEvent, UserDeletedEvent } from "@src/domain/events";
import { UserId } from "@src/domain/models";

/**
 * Represents the properties required to create a user.
 * @extends {EntityCreateProps}
 */
export interface CreateUserProps extends EntityCreateProps {
    name: string;
}
/**
 * Represents the properties required to update a user.
 * @extends {EntityUpdateProps}
 */
export interface UpdateUserProps extends EntityUpdateProps<UserId> {
    name: string;
}
/**
 * Represents a user aggregate.
 * @extends AggregateRoot
 */
export class User extends AggregateRoot<UserId> {
    /**
     * Initializes a new instance of a user.
     * @param {UserId} id - An optional identifier of the user.
     * @param {string} name - The name of the user.
     */
    private constructor(private name: string, id?: UserId) {
        super(id ?? UserId.create());
        if (!id) {
          this.applyChange(new UserCreatedEvent(this.id, this));
        }
    }

    /**
     * Gets the name of the user.
     * @returns {string} - The name of the user.
     */
    getName(): string {
        return this.name;
    }

    /**
     * Update the user.
     * @param {UpdateUserProps} props - The properties to update.
     * @returns void
     */
    update(props: UpdateUserProps): void {
        // TODO: validate props
        this.name = props.name;
        this.applyChange(new UserUpdatedEvent(this.id, this));
    }

    delete(): void {
        this.applyChange(new UserDeletedEvent(this.id, this));
    }

    static create(props: CreateUserProps, id?: UserId): User {
        const user = new User(
          props.name,
          id
        );
        return user;
    }
}    
"
`;

exports[`Aggregate Should render the trackable aggregate component 1`] = `
"import { TrackableAggregateRoot, EntityCreateProps, EntityUpdateProps, UtcDate } from "@typescript-ddd/core";
import { UserCreatedEvent, UserUpdatedEvent, UserDeletedEvent } from "@src/domain/events";
import { UserId } from "@src/domain/models";

/**
 * Represents the properties required to create a user.
 * @extends {EntityCreateProps}
 */
export interface CreateUserProps extends EntityCreateProps {
    name: string;
    createdAtUtc?: UtcDate;
    updatedAtUtc?: UtcDate;
}
/**
 * Represents the properties required to update a user.
 * @extends {EntityUpdateProps}
 */
export interface UpdateUserProps extends EntityUpdateProps<UserId> {
    name: string;
}
/**
 * Represents a user aggregate.
 * @extends TrackableAggregateRoot
 */
export class User extends TrackableAggregateRoot<UserId> {
    /**
     * Initializes a new instance of a user.
     * @param {UserId} id - An optional identifier of the user.
     * @param {string} name - The name of the user.
     * @param {UtcDate} createdAtUtc - An optional creation date.
     * @param {UtcDate} updatedAtUtc - An optional last update date.
     */
    private constructor(private name: string, createdAtUtc?: UtcDate, updatedAtUtc?: UtcDate, id?: UserId) {
        super(id ?? UserId.create(), createdAtUtc, updatedAtUtc);
        if (!id) {
          this.applyChange(new UserCreatedEvent(this.id, this));
        }
    }

    /**
     * Gets the name of the user.
     * @returns {string} - The name of the user.
     */
    getName(): string {
        return this.name;
    }

    /**
     * Update the user.
     * @param {UpdateUserProps} props - The properties to update.
     * @returns void
     */
    update(props: UpdateUserProps): void {
        // TODO: validate props
        this.name = props.name;
        this.updatedAtUtc = UtcDate.now();
        this.applyChange(new UserUpdatedEvent(this.id, this));
    }

    delete(): void {
        this.applyChange(new UserDeletedEvent(this.id, this));
    }

    static create(props: CreateUserProps, id?: UserId): User {
        const user = new User(
          props.name,
          id
        );
        return user;
    }
}    
"
`;
